cli/azd/extensions/azure.coding-agent/internal/cmd/coding_agent_config.go

```diff
@@ -0,0 +1,527 @@
+		return err
+	}
+
+	varsToSet := map[string]string{
+		"AZURE_CLIENT_ID":       authConfig.ClientId,
+		"AZURE_TENANT_ID":       authConfig.TenantId,
+		"AZURE_SUBSCRIPTION_ID": authConfig.SubscriptionId,
+	}
+
+	for name, value := range varsToSet {
+		if err := cli.SetVariable(ctx, cmdFlags.RepoSlug, name, value, &github.SetVariableOptions{
+			Environment: copilotEnv,
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type credentialProviderAdapter struct {
+	tokenCred azcore.TokenCredential
+}
+
+func (cp *credentialProviderAdapter) CredentialForSubscription(ctx context.Context, subscriptionId string) (azcore.TokenCredential, error) {
+	return cp.tokenCred, nil
+}
+
+// CreateFederatedCredential creates a federated credential (allowing Copilot to authenticate and use Azure)
+func CreateFederatedCredential(ctx context.Context,
+	msiService azd_armmsi.ArmMsiService,
+	repoSlug string,
+	copilotEnvName string,
+	subscriptionId string,
+	msiResourceID string) error {
+	// copied from azd's github_provider.go
+	const (
+		federatedIdentityIssuer   = "https://token.actions.githubusercontent.com"
+		federatedIdentityAudience = "api://AzureADTokenExchange"
+	)
+
+	credentialSafeName := strings.ReplaceAll(repoSlug, "/", "-")
+
+	armFedCreds := []rm_armmsi.FederatedIdentityCredential{
+		{
+			Name: to.Ptr(url.PathEscape(fmt.Sprintf("%s-copilot-env", credentialSafeName))),
+			Properties: &rm_armmsi.FederatedIdentityCredentialProperties{
+				Subject:   to.Ptr(fmt.Sprintf("repo:%s:environment:%s", repoSlug, copilotEnvName)),
+				Issuer:    to.Ptr(federatedIdentityIssuer),
+				Audiences: []*string{to.Ptr(federatedIdentityAudience)},
+			},
+		},
+	}
+
+	if _, err := msiService.ApplyFederatedCredentials(ctx, subscriptionId, msiResourceID, armFedCreds); err != nil {
+		return fmt.Errorf("failed to create federated credentials: %w", err)
+	}
+
+	return nil
+}
+
+func GetExistingMSIByResourceID(ctx context.Context,
+	msiService azd_armmsi.ArmMsiService,
+	managedIdentityResourceID string) (*authConfiguration, error) {
+	id, err := msiService.GetUserIdentity(ctx, managedIdentityResourceID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get user identity for resource ID %s: %w", managedIdentityResourceID, err)
+	}
+
+	return &authConfiguration{
+		ClientId:       *id.Properties.ClientID,
+		SubscriptionId: *id.Properties.TenantID,
+		TenantId:       *id.Properties.TenantID,
+		ResourceID:     *id.ID,
+	}, nil
+}
+
+// PickOrCreateMSI walks the user through creating an MSI
+func PickOrCreateMSI(ctx context.Context,
+	prompter azdext.PromptServiceClient,
+	msiService azd_armmsi.ArmMsiService,
+	entraIDService entraid.EntraIdService,
+	resourceService interface {
+		CreateOrUpdate(ctx context.Context, resourceGroupName string, parameters armresources.ResourceGroup, options *armresources.ResourceGroupsClientCreateOrUpdateOptions) (armresources.ResourceGroupsClientCreateOrUpdateResponse, error)
```

Comment:
Would it make more sense to define this not inline? I'm seeing this twice

Reply:
Absolutely! I didn't even realize I'd done it.